dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +241,260 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
+                }
 
-            auto registrationActivators{ details.Activators() };
+                return false;
+            }) };
 
-            auto isBackgroundTaskFlagSet{ WI_IsAnyFlagSet(registrationActivators, PushNotificationRegistrationActivators::PushTrigger | PushNotificationRegistrationActivators::ComActivator) };
-            THROW_HR_IF(E_INVALIDARG, isBackgroundTaskFlagSet && !IsActivatorSupported(registrationActivators));
+        return isTaskRegistered;
+    }
 
-            auto isProtocolActivatorSet{ WI_IsFlagSet(registrationActivators, PushNotificationRegistrationActivators::ProtocolActivator) };
-            THROW_HR_IF(E_INVALIDARG, isProtocolActivatorSet && !IsActivatorSupported(registrationActivators));
+    void PushNotificationManager::Register()
+    {
+        static bool registering{ false };
+        {
+            auto lock{ m_lock.lock_exclusive() };
+            THROW_HR_IF_MSG(E_FAIL, registering, "Registration is in progress!");
+            registering = true;
+        }
 
-            if (isProtocolActivatorSet)
+        try
+        {
+            if (PushNotificationHelpers::IsPackagedAppScenario())
             {
+                auto scopeExitToCleanRegistrations{ wil::scope_exit([&]()
                 {
-                    auto lock = s_activatorInfoLock.lock_shared();
-                    THROW_HR_IF(E_INVALIDARG, s_protocolRegistration);
-                }
+                    m_comActivatorRegistration.reset();
 
-                // AppId generated by PushNotificationsLongRunningTask singleton
-                wil::unique_cotaskmem_string unpackagedAppUserModelId;
-                RegisterUnpackagedApplicationHelper(GUID_NULL, unpackagedAppUserModelId); // create default registration for app
+                    // Clean the task registration only if it was created during this call
+                    if (m_pushTriggerRegistration)
+                    {
+                        m_pushTriggerRegistration.Unregister(true);
+                        m_pushTriggerRegistration = nullptr;
+                    }
 
-                {
-                    auto lock = s_activatorInfoLock.lock_exclusive();
-                    s_protocolRegistration = true;
+                    auto appUserModelId{ PushNotificationHelpers::GetAppUserModelId() };
+                    PushNotifications_UnregisterNotificationSinkForFullTrustApplication(appUserModelId.get());
                 }
-            }
-
-            BackgroundTaskBuilder builder{ nullptr };
-
-            if (WI_IsFlagSet(registrationActivators, PushNotificationRegistrationActivators::PushTrigger) && PushNotificationHelpers::IsPackagedAppScenario())
-            {
-                GUID taskClsid = details.TaskClsid();
-                THROW_HR_IF(E_INVALIDARG, taskClsid == GUID_NULL);
+                )};
 
+                // Register a PushTrigger for BI to activate the application through COM
+                if (!m_pushTriggerRegistration)
```

Comment:
Remove this if check. There should be a check earlier in IsPackagedScenario() to ensure that this isn't already set and throw if it is.